### ISSUE TITLE
Conv Ouput Geometry issues

### ISSUE THREAD

**Issue by crutcher** [tracel-ai/burn#3705]
**Describe the bug**
This is on 0.18.0; haven't verified on main yet.

1. Some conv-family operations misshandle geometry for some backends.
2. NdArray is broken in the below case for `conv1d`, but `conv2d` has the same issue.
3. There is no standard geometry output contract that conv-family operations are compared to.
4. There is no *documentation* for that contract, just a vague expectation that it matches `pytorch`.

**To Reproduce**
```rust
    #[test]
    fn test_conv1d_edge_case_wgpu() {
        // This passes.
        check_conv1d_edge_case::<burn::backend::Wgpu>();
    }

    #[test]
    fn test_conv1d_edge_case_ndarray() {
        // This dies.
        check_conv1d_edge_case::<burn::backend::NdArray>();
    }

    #[allow(unused)]
    fn check_conv1d_edge_case<B: Backend>() {
        let device = Default::default();

        let input_shape = 3;
        let kernel_shape = 3;
        let stride = 1;
        let dilation = 3;
        let padding = 2;

        let batch_size = 1;
        let in_channels = 4;
        let out_channels = 8;

        let input: Tensor<B, 3> = Tensor::ones([batch_size, in_channels, input_shape], &device);
        let weight: Tensor<B, 3> = Tensor::ones([out_channels, in_channels, kernel_shape], &device);

        // See: https://docs.pytorch.org/docs/stable/generated/torch.nn.Conv1d.html
        // out_shape = floor( ((in_shape + 2*padding - dilation*(kernel_shape-1) - 1) / stride) + 1 )
        //   = floor( ((3 + 2*2 - 3*(3-1) - 1) / 1) + 1 )
        //   = floor( (3 + 4 - 6 - 1) + 1 )
        //   = floor( (7 - 7) + 1 )
        //   = floor( 1 )
        let output = conv1d(
            input,
            weight,
            None,
            ConvOptions {
                stride: [stride],
                padding: [padding],
                dilation: [dilation],
                groups: 1,
            },
        );
        assert_eq!(output.shape().dims(), [batch_size, out_channels, 1])
    }
```

**Comment by crutcher** [comment]
@wingertge I think you have the most context?

**Comment by crutcher** [comment]
I want something like this, for all the conv-like interfaces:
```rust
pub fn conv1d_output_shape(
    input_shape: usize,
    kernel_shape: usize,
    stride: usize,
    dilation: usize,
    padding: usize,
    use_float: bool,
) -> anyhow::Result<usize>;
```

**Comment by crutcher** [comment]
I'll try and firm up a contract for some of these, and do a more complete cross-check against the existing backends on main

**Comment by wingertge** [comment]
> [@wingertge](https://github.com/wingertge) I think you have the most context?

I actually never touched the shape stuff, I just left the setup code in place and only touched the actual algorithm. I agree the function should return the entire shape, instead of the weird half-solution we have right now that only returns the spatial dimensions one by one. I just always minimized changes so never touched that part of the code.

**Comment by crutcher** [comment]
```rust
//! # Convolution Shape Utilities
//!
//! Utilities for computing the output shape of convolution operations.

/// Predict the output shape of a 1D convolution operation.
///
/// ```text
/// out_shape = floor( ((in_shape + 2*padding - dilation*(kernel_shape-1) - 1) / stride) + 1 )
/// ```
///
/// # Reference
///
/// - [conv_arithmetic diagram](https://github.com/vdumoulin/conv_arithmetic/blob/master/README.md)
///   visual explanations of these convolution parameters.
/// - [pytorch conv1d](https://docs.pytorch.org/docs/stable/generated/torch.nn.Conv1d.html)
///
/// # Arguments
///
/// - `input_shape`: The input dimension shape, must be > 0.
/// - `kernel_shape`: The kernel shape, must be > 0.
/// - `stride`: The stride of the convolution, must be > 0.
/// - `dilation`: The dilation of the convolution, must be > 0.
/// - `padding`: The padding of the convolution, added evenly to all sides of the input.
///
/// # Returns
///
/// An `Option<usize>` representing the output shape; or `None` for <= 0.
pub fn conv1d_output_shape(
    input_shape: usize,
    kernel_shape: usize,
    stride: usize,
    dilation: usize,
    padding: usize,
) -> Option<usize> {
    assert!(input_shape > 0);
    assert!(kernel_shape > 0);
    assert!(stride > 0);
    assert!(dilation > 0);

    let effective_shape = input_shape + 2 * padding;
    let pos = effective_shape + stride;
    let kernel_width = 1 + dilation * (kernel_shape - 1);

    if pos < kernel_width {
        return None;
    }
    let x = (pos - kernel_width) / stride;
    if x < 1 { None } else { Some(x) }
}

/// Predict the output shape of a D convolution operation.
///
/// This is the generalization of [`conv1d_output_shape`] to D dimensions.
///
/// # Arguments
///
/// - `input_shape`: The input dimension shape, each dim must be > 0.
/// - `kernel_shape`: The kernel shape, each dim must be > 0.
/// - `stride`: The stride of the convolution, each dim must be > 0.
/// - `dilation`: The dilation of the convolution, each dim must be > 0.
/// - `padding`: The padding of the convolution, added evenly to all sides of the input.
///
/// # Returns
///
/// An `Option<[usize; D]>` representing the output shape; or `None` for <= 0.
pub fn conv_output_shape<const D: usize>(
    input_shape: [usize; D],
    kernel_shape: [usize; D],
    stride: [usize; D],
    dilation: [usize; D],
    padding: [usize; D],
) -> Option<[usize; D]> {
    let mut output_shape = input_shape;
    for i in 0..D {
        output_shape[i] = conv1d_output_shape(
            input_shape[i],
            kernel_shape[i],
            stride[i],
            dilation[i],
            padding[i],
        )?;
    }
    Some(output_shape)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_conv1d_output_shape() {
        pub fn conv1d_output_shape_reference(
            input_shape: usize,
            kernel_shape: usize,
            stride: usize,
            dilation: usize,
            padding: usize,
        ) -> Option<usize> {
            let input_shape = input_shape as f64;
            let kernel_shape = kernel_shape as f64;
            let stride = stride as f64;
            let dilation = dilation as f64;
            let padding = padding as f64;

            let effective_shape = input_shape + 2.0 * padding;
            let kernel_width = 1.0 + dilation * (kernel_shape - 1.0);

            let x = (((effective_shape - kernel_width) / stride) + 1.0).floor();
            if x < 1.0 { None } else { Some(x as usize) }
        }

        for input_shape in 1..10 {
            for stride in 1..3 {
                for kernel_shape in 1..4 {
                    for dilation in 1..2 {
                        for padding in 0..10 {
                            assert_eq!(
                                conv1d_output_shape(
                                    input_shape,
                                    kernel_shape,
                                    stride,
                                    dilation,
                                    padding,
                                ),
                                conv1d_output_shape_reference(
                                    input_shape,
                                    kernel_shape,
                                    stride,
                                    dilation,
                                    padding,
                                )
                            )
                        }
                    }
                }
            }
        }
    }
}
```

**Comment by antimora** [comment]
Retested on main: your conv1d edge case returns `[1, 8, 1]` on flex, ndarray and Metal. Closing.
